wasm/src/graphs.c

```diff
@@ -0,0 +1,87 @@
+#include "graphs.h"
+
+static void __builtin_graph_reachable(opa_value *edges, opa_array_t *queue, opa_set_t *reached)
+{
+    switch (opa_value_type(edges))
+    {
+        case OPA_SET:
+        {
+            opa_set_t *x = opa_cast_set(edges);
+
+            for (int i = 0; i < x->n; i++)
+            {
+                opa_set_elem_t *elem = x->buckets[i];
+
+                while (elem != NULL)
+                {
+                    if (reached == NULL || opa_set_get(reached, elem->v) == NULL)
+                    {
+                        opa_array_append(queue, elem->v);
+                    }
+                    elem = elem->next;
+                }
+            }
+        }  
+        case OPA_ARRAY: 
+        {
+            opa_array_t *y = opa_cast_array(edges);
+
+            for (int i = 0; i < y->len; i++)
+            {
+                opa_value *elem = y->elems[i].v;
+
+                if (reached == NULL || opa_set_get(reached, elem) == NULL)
+                {
+                    opa_array_append(queue, elem);
+                }
+            }
+        }
+    }
+}
+
+opa_value *builtin_graph_reachable(opa_value *graph, opa_value *initial)
+{
+
+     if (opa_value_type(graph) != OPA_OBJECT)
+    {
+        return opa_set();
```

Comment:
This is different from the behaviour of the topdown implementation, isn't it? I'd get an error there if passing something that's not an object in operand 1.

Reply:
The error you'll see is returned by the compiler's type checker. The topdown implementation returns an empty set if the first argument is not an object.